reducer/src/main/java/com/graphicsfuzz/reducer/glslreducers/EliminateInjectionMacrosVisitor.java

```diff
@@ -16,128 +16,84 @@
+          || MacroNames.isIfWrapperFalse(expr)
+          || MacroNames.isIfWrapperTrue(expr)
+      ) {
+        parent.replaceChild(child,
+            addParenthesesIfNecessary(parent, ((FunctionCallExpr) child).getChild(0)));
+      }
     }
   }
 
-  private void cleanupLoop(LoopStmt loopStmt) {
-    if (MacroNames.isLoopWrapper(loopStmt.getCondition())) {
-      loopStmt.setCondition(loopStmt.getCondition().getChild(0));
+  private IAstNode addParenthesesIfNecessary(IAstNode parent, Expr child) {
+    if (ifParenthesesNecessary(parent, child)) {
+      return new ParenExpr(child);
     }
+    return child;
   }
 
-  @Override
-  public void visitDoStmt(DoStmt doStmt) {
-    super.visitDoStmt(doStmt);
-    cleanupLoop(doStmt);
-  }
+  boolean ifParenthesesNecessary(IAstNode parent, Expr child) {
+    if (child instanceof ConstantExpr
+        || child instanceof ParenExpr
+        || child instanceof VariableIdentifierExpr
+        || child instanceof FunctionCallExpr) {
+      // Parentheses is unnecessary in cases such as _GLF_FUNCTION(1),
+      // _GLF_FUNCTION((1)), _GLF_FUNCTION(a), _GLF_FUNCTION(sin(a)).
+      return false;
+    }
 
-  @Override
-  public void visitForStmt(ForStmt forStmt) {
-    super.visitForStmt(forStmt);
-    cleanupLoop(forStmt);
-  }
+    if (!(parent instanceof Expr)) {
+      // No parentheses needed if the parent is not an expression,
+      // for example, int x = _GLF_FUNCTION(a + b).
+      return false;
+    }
 
-  private void cleanUpMacros(Expr parent) {
-    for (int i = 0; i < parent.getNumChildren(); i++) {
-      Expr child = parent.getChild(i);
-      // Note: it would be redundant to have a special function reduction opportunity
-      // be classed also as a fuzzed expression reduction opportunity
-      if (MacroNames.isIdentity(child)
-          || MacroNames.isZero(child)
-          || MacroNames.isOne(child)
-          || MacroNames.isFalse(child)
-          || MacroNames.isTrue(child)) {
-        replaceChildWithGrandchild(parent, i, 1);
-      } else if (MacroNames.isFuzzed(child)
-          || MacroNames.isDeadByConstruction(child)) {
-        replaceChildWithGrandchild(parent, i, 0);
-      }
+    if (parent instanceof ParenExpr) {
+      // If parent is parentheses, adding a new parentheses would be redundant,
+      // e.g. (_GLF_FUNCTION(a + b)).
+      return false;
     }
-  }
 
-  @Override
-  public void visitScalarInitializer(ScalarInitializer scalarInitializer) {
-    super.visitScalarInitializer(scalarInitializer);
-    if (MacroNames.isFuzzed(scalarInitializer.getExpr())) {
-      scalarInitializer.setExpr(scalarInitializer.getExpr().getChild(0));
+    if (parent instanceof FunctionCallExpr) {
+      if (child instanceof BinaryExpr) {
+        // The binary operator must not be the comma operator
+        // as it cannot appear directly as a macro argument.
+        assert ((BinaryExpr) child).getOp() != BinOp.COMMA;
+        return true;
```

Comment:
I do not understand why you return true here.  Let's say the parent is a function call expression, e.g. foo, and the child is a binary expression, e.g. a + b.

We thus have something like:

foo(..., a + b, ...)

there is no need to put parentheses around a + b.

I suggest replacing this whole conditional with:

assert (!(child instanceof BinaryExpr) || ((BinaryExpr) child).getOp() != BinOp.COMMA);

and keeping the nice comment about the reason for the comma operator.

But I believe false can be returned whenever parent is a FunctionCallExpr.

Reply:
Can you add an additional test that for the above case, e.g. something that checks that:

foo(_GLF_IDENTITY(1 + 2, 1 + 2))

should get simplified to:

foo(1 + 2)
?

You should find that the test fails with your current code, and after you address the above comment it should pass.

Reply:
Got it! thank you for a very clear explanation.